sdk/cosmosdb/cosmos/src/queryExecutionContext/parallelQueryExecutionContextBase.ts

```diff
@@ -539,4 +504,115 @@ export abstract class ParallelQueryExecutionContextBase implements ExecutionCont
       options
     );
   }
+
+  private async _createDocumentProducersAndFillUpPriorityQueue(
+    operationOptions?: QueryOperationOptions,
+    ruConsumedManager?: RUConsumedManager
+  ): Promise<void> {
+    try {
+      const targetPartitionRanges = await this._onTargetPartitionRanges();
+      const maxDegreeOfParallelism =
+        this.options.maxDegreeOfParallelism === undefined || this.options.maxDegreeOfParallelism < 1
+          ? targetPartitionRanges.length
+          : Math.min(this.options.maxDegreeOfParallelism, targetPartitionRanges.length);
+
+      logger.info(
+        "Query starting against " +
+          targetPartitionRanges.length +
+          " ranges with parallelism of " +
+          maxDegreeOfParallelism
+      );
+
+      let filteredPartitionKeyRanges = [];
+      // The document producers generated from filteredPartitionKeyRanges
+      const targetPartitionQueryExecutionContextList: DocumentProducer[] = [];
+
+      if (this.requestContinuation) {
+        throw new Error("Continuation tokens are not yet supported for cross partition queries");
+      } else {
+        filteredPartitionKeyRanges = targetPartitionRanges;
+      }
+
+      // Create one documentProducer for each partitionTargetRange
+      filteredPartitionKeyRanges.forEach((partitionTargetRange: any) => {
+        // TODO: any partitionTargetRange
+        // no async callback
+        targetPartitionQueryExecutionContextList.push(
+          this._createTargetPartitionQueryExecutionContext(partitionTargetRange)
+        );
+      });
+
+      // Fill up our priority queue with documentProducers
+      let inProgressPromises: Promise<void>[] = [];
+      for (const documentProducer of targetPartitionQueryExecutionContextList) {
+        // Don't enqueue any new promise if RU cap exceeded
+        if (this.ruCapExceededError) {
+          break;
+        }
+        const promise: Promise<void> = this._processAndEnqueueDocumentProducer(
+          documentProducer,
+          operationOptions,
+          ruConsumedManager
+        );
+        inProgressPromises.push(promise);
+
+        // Limit concurrent executions
+        if (inProgressPromises.length === maxDegreeOfParallelism) {
+          await Promise.all(inProgressPromises);
+          inProgressPromises = [];
+        }
+      }
+      // Wait for all promises to complete
+      await Promise.all(inProgressPromises);
+      if (this.err) {
```

Comment:
We need to use `Promise.allSettled()` to wait for all document producers as buffered items might not have filled for other document producers when first ruCapExceededError is thrown. 
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled